metrics/prometheus/prometheus_test.go

```diff
@@ -8,80 +8,121 @@ import (
 
 	timer := metrics.NewTimer()
-	defer timer.Stop()
+	t.Cleanup(timer.Stop)
 	timer.Update(20 * time.Millisecond)
 	timer.Update(21 * time.Millisecond)
 	timer.Update(22 * time.Millisecond)
 	timer.Update(120 * time.Millisecond)
 	timer.Update(23 * time.Millisecond)
 	timer.Update(24 * time.Millisecond)
-
-	err = registry.Register("test/timer", timer)
-	assert.NoError(t, err)
+	register(t, "test/timer", timer)
 
 	resettingTimer := metrics.NewResettingTimer()
-	resettingTimer.Update(10 * time.Millisecond)
-	resettingTimer.Update(11 * time.Millisecond)
-	resettingTimer.Update(12 * time.Millisecond)
-	resettingTimer.Update(120 * time.Millisecond)
-	resettingTimer.Update(13 * time.Millisecond)
-	resettingTimer.Update(14 * time.Millisecond)
-
-	err = registry.Register("test/resetting_timer", resettingTimer)
-	assert.NoError(t, err)
-
-	err = registry.Register("test/resetting_timer_snapshot", resettingTimer.Snapshot())
-	assert.NoError(t, err)
+	register(t, "test/resetting_timer", resettingTimer)
+	resettingTimer.Update(time.Second) // must be after register call
 
 	emptyResettingTimer := metrics.NewResettingTimer()
-
-	err = registry.Register("test/empty_resetting_timer", emptyResettingTimer)
-	assert.NoError(t, err)
-
-	err = registry.Register("test/empty_resetting_timer_snapshot", emptyResettingTimer.Snapshot())
-	assert.NoError(t, err)
-
-	g := Gatherer(registry)
-
-	_, err = g.Gather()
-	assert.NoError(t, err)
+	register(t, "test/empty_resetting_timer", emptyResettingTimer)
+
+	emptyResettingTimer.Update(time.Second) // no effect because of snapshot below
+	register(t, "test/empty_resetting_timer_snapshot", emptyResettingTimer.Snapshot())
+
+	// Nil metrics
+	metrics.Enabled = false
+	nilCounter := metrics.NewCounter()
+	register(t, "nil/counter", nilCounter)
+	nilCounterFloat64 := metrics.NewCounterFloat64()
+	register(t, "nil/counter_float64", nilCounterFloat64)
+	nilEWMA := &metrics.NilEWMA{}
+	register(t, "nil/ewma", nilEWMA)
+	nilGauge := metrics.NewGauge()
+	register(t, "nil/gauge", nilGauge)
+	nilGaugeFloat64 := metrics.NewGaugeFloat64()
+	register(t, "nil/gauge_float64", nilGaugeFloat64)
+	nilGaugeInfo := metrics.NewGaugeInfo()
+	register(t, "nil/gauge_info", nilGaugeInfo)
+	nilHealthcheck := metrics.NewHealthcheck(nil)
+	register(t, "nil/healthcheck", nilHealthcheck)
+	nilHistogram := metrics.NewHistogram(nil)
+	register(t, "nil/histogram", nilHistogram)
+	nilMeter := metrics.NewMeter()
+	register(t, "nil/meter", nilMeter)
+	nilResettingTimer := metrics.NewResettingTimer()
+	register(t, "nil/resetting_timer", nilResettingTimer)
+	nilSample := metrics.NewUniformSample(1028)
+	register(t, "nil/sample", nilSample)
+	nilTimer := metrics.NewTimer()
+	register(t, "nil/timer", nilTimer)
+
+	metrics.Enabled = true
+
+	gatherer := NewGatherer(registry)
+
+	families, err := gatherer.Gather()
+	require.NoError(t, err)
+	familyStrings := make([]string, len(families))
+	for i := range families {
+		familyStrings[i] = families[i].String()
+	}
```

Comment:
(fyi) This is where the `cmp` package shines. In this case you'd probably want a [Comparer](https://pkg.go.dev/github.com/google/go-cmp/cmp#Comparer). Although a `Transformer` would do the same thing, it's a means (converting to a string) to an end (checking equality).

Reply:
1. Comparer would need both want and got to be of the same type right? Which is not the case here, and, if it would be, we could just use `assert.Equal`. I have my `want` as strings since it's easier to read/more compact than having it as `*dto.MetricFamily` structs.
2. I tried using `cmp.Transformer`, using:

    ```go
    transformer := cmp.Transformer("family_string", func(f *dto.MetricFamily) string {
        return f.String()
    })
    assert.Empty(t, cmp.Diff(want, families, transformer))
    ```

    but it doesn't seem to convert the `families` to a string slice? 🤔

Reply:
You may need to define the transformer as `func([]*dto.MetricFamily) []string {`. If you want to go ahead with using it then DM me and I'll help figure it out otherwise let's just leave it.